File: wildcard-matching.java
// https://leetcode.com/problems/wildcard-matching/description/
// DFS : compare 1 by 1 (string & pattern)
// 1. When '*': pattern++
// 2. When '?' or matched: string++ & pattern++
// 3. When ends: if string is remaining, try another!
//    otherwise break
// Time: O(3^n), Space: O(1)
class Solution {
    public boolean isMatch(String s, String p) {
        return dfs(s,p,0,0)>1;        
    }
    private int dfs(String s, String p, int si, int pi){
        int sl = s.length(), pl = p.length();
        if(si==sl && pi==pl) return 2; //matched
        if(si==sl && p.charAt(pi)!='*') return 0; //unmatched with all s
        if(pi==pl) return 1; //unmatched without all s
        if(p.charAt(pi)=='*'){
            if(pi+1<pl && p.charAt(pi+1)=='*') return dfs(s,p,si,pi+1); //skip **
            for(int i=0; i<=(sl-si); i++){
                int r = dfs(s,p,si+i,pi+1);
                if(r==0 || r==2) return r;
            }
        }
        if(p.charAt(pi)=='?' || s.charAt(si)==p.charAt(pi)) return dfs(s,p,si+1,pi+1);
        return 1;
    }
}

File: wildcard-matching.py
# https://leetcode.com/problems/wildcard-matching/description/
class Solution:
    def isMatch(self, s, p):
        return self._dfs(s, p, 0, 0) > 1
    def _dfs(self, s, p, si, pi):
        if si==len(s):
            if pi==len(p): return 2 #matched
            if p[pi]!='*': return 0 #unmatched w/ all s
        if pi==len(p): return 1 #unmatched w/o all s (try more)
        if p[pi]=='*':
            if pi+1<len(p) and p[pi+1]=='*': return self._dfs(s, p, si, pi+1) #skip **
            for i in range(len(s)-si+1):
                r = self._dfs(s, p, si+i, pi+1) #try all possibles
                if r==0 or r==2: return r #if matched or unmatched w/ all s, stop
        if p[pi]=='?' or s[si]==p[pi]: return self._dfs(s, p, si+1, pi+1) #go 1 step
        return 1 #unmatched w/o all s (try more)
        
